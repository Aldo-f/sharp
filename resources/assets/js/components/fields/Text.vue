<template>
    <input :type="inputType" class="form-control"
           :value="value"  :placeholder="placeholder" :disabled="disabled"
           @input="handleInput"
    >
</template>

<script>
    export default {
        name:'SharpText',

        inject: ['value'],

        props: {
            value: {
                type: String,
            },
            placeholder: String,
            disabled: Boolean,
            inputType: {
                type: String,
                default: 'text'
            }
        },
        data() {
            return {}
        },
        methods: {
            handleInput(e) {
                this.$emit('input',e.target.value);
            }
        }
    }

</script>