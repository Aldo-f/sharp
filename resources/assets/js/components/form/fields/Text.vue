<template>
    <input :type="inputType" class="SharpText"
           :value="value"  :placeholder="placeholder" :disabled="disabled"
           @input="handleInput" ref="input"
    >
</template>

<script>
    import Vue from 'vue';
    import { Focusable } from '../../../mixins';

    export default {
        name:'SharpText',

        mixins: [Focusable],

        props: {
            value: [String, Number],

            placeholder: String,
            disabled: Boolean,
            inputType:  {
                type:String,
                default:'text'
            },
        },
        data() {
            return {}
        },
        methods: {
            handleInput(e) {
                this.$emit('input',e.target.value);
            }
        },

        mounted() {
            this.setFocusable(this.$refs.input);
        }
    }
</script>