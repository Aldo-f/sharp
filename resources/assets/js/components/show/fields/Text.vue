<template>
    <div class="ShowTextField">
        <template v-if="label">
            <span class="ShowTextField__label">{{ label }}</span>
        </template>
        <div class="ShowTextField__content" v-html="currentContent" ref="content"></div>
        <template v-if="hasCollapsed">
            <div class="mt-2">
                <a href="#" class="text-decoration-none text-lowercase" @click.prevent="handleToggleClicked">
                    <template v-if="expanded">
                        - {{ l('show.text.show_less') }}
                    </template>
                    <template v-else>
                        + {{ l('show.text.show_more') }}
                    </template>
                </a>
            </div>
        </template>
    </div>
</template>

<script>
    import { Localization } from '../../../mixins';
    import clip from 'text-clipper';

    function stripTags(html) {
        const el = document.createElement('div');
        el.innerHTML = html;
        return el.textContent;
    }

    function truncateToWords(text, count) {
        const matches = [...text.matchAll(/\S+\s*/g)];
        return matches.length > count
            ? matches.slice(0, count).map(match => match[0]).join('')
            : text;
    }

    export default {
        mixins: [Localization],
        props: {
            value: String,
            collapseToWordCount: Number,
            label: String,
        },
        data() {
            return {
                expanded: false,
            }
        },
        computed: {
            currentContent() {
                if(this.hasCollapsed && !this.expanded) {
                    return this.collapsedContent;
                }
                return this.value;
            },
            hasCollapsed() {
                return !!this.collapsedContent;
            },
            collapsedContent() {
                if(!this.collapseToWordCount || !this.value) {
                    return null;
                }
                const html = this.value.trim();
                const text = stripTags(html);
                const truncated = truncateToWords(text, this.collapseToWordCount);
                return truncated.length < text.length
                    ? clip(html, truncated.length + 2, { html: true })
                    : null;
            },
        },
        methods: {
            handleToggleClicked() {
                this.expanded = !this.expanded;
            },
        },
    }
</script>