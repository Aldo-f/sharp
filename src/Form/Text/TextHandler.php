<?php

namespace Code16\Sharp\Form\Text;

class TextHandler
{

}