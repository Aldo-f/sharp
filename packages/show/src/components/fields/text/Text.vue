<template>
    <FieldLayout class="ShowTextField" :class="classes" :label="label">
        <template v-if="html">
            <TextRenderer
                class="ShowTextField__content"
                :content="currentContent"
            />
        </template>
        <template v-else>
            <div class="ShowTextField__content">{{ currentContent }}</div>
        </template>
        <template v-if="hasCollapsed">
            <div class="mt-2">
                <a href="#" class="ShowTextField__more" @click.prevent="handleToggleClicked">
                    <template v-if="expanded">
                        - {{ l('show.text.show_less') }}
                    </template>
                    <template v-else>
                        + {{ l('show.text.show_more') }}
                    </template>
                </a>
            </div>
        </template>
    </FieldLayout>
</template>

<script>
    import { Localization } from 'sharp/mixins';
    import clip from 'text-clipper';
    import { syncVisibility } from "../../../util/fields/visiblity";
    import { truncateToWords, stripTags } from "../../../util/fields/text";
    import FieldLayout from "../../FieldLayout";
    import TextRenderer from "./TextRenderer";

    export default {
        mixins: [Localization],
        components: {
            FieldLayout,
            TextRenderer,
        },
        props: {
            value: String,
            collapseToWordCount: Number,
            label: String,
            emptyVisible: Boolean,
            html: Boolean,
        },
        data() {
            return {
                expanded: false,
            }
        },
        computed: {
            classes() {
                return {
                    'ShowTextField--html': this.html,
                }
            },
            currentContent() {
                if(!this.value) {
                    return null;
                }
                if(this.hasCollapsed && !this.expanded) {
                    return this.collapsedContent;
                }
                if(!this.html) {
                    return stripTags(this.value).trim();
                }
                return this.value;
            },
            hasCollapsed() {
                return !!this.collapsedContent;
            },
            collapsedContent() {
                if(!this.collapseToWordCount || !this.value) {
                    return null;
                }
                const value = this.value.trim();
                const text = stripTags(value);
                const content = this.html ? value : text;
                const truncated = truncateToWords(text, this.collapseToWordCount);
                return truncated.length < text.length
                    ? clip(content, truncated.length + 2, { html: this.html })
                    : null;
            },
            isVisible() {
                return !!this.value || this.emptyVisible;
            },
        },
        methods: {
            handleToggleClicked() {
                this.expanded = !this.expanded;
            },
        },
        created() {
            syncVisibility(this, () => this.isVisible);
        },
    }
</script>
