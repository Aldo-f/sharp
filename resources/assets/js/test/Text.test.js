import Vue from 'vue/dist/vue';
import Text from '../components/form/fields/Text.vue';



describe('text-fields', () => {
    Vue.component('sharp-text', Text);

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-text placeholder="Entrez du texte" value="AAA"></sharp-text>
            </div>
        `
    });

    it('can mount Text field', async () => {
        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });
});

async function createVm() {
    const vm = new Vue({
        el: '#app',
    });

    await Vue.nextTick();

    return vm.$children[0];
}