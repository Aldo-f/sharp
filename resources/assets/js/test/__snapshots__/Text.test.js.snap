// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text-fields can mount Text field 1`] = `
"
            <div id=\\"app\\"><input type=\\"text\\" placeholder=\\"Entrez du texte\\" class=\\"form-control\\"></div>
        "
`;
